cnf-certification-test/networking/suite.go

```diff
@@ -67,8 +80,103 @@ var _ = ginkgo.Describe(common.NetworkingTestKey, func() {
+	ginkgo.It(testID, ginkgo.Label(testID), func() {
+		testListenAndDeclared(&env)
+	})
 })
 
+func parseListening(res string, listeningPorts map[key]string) {
+	var k key
+	lines := strings.Split(res, "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if !strings.Contains(line, "LISTEN") {
+			continue
+		}
+		if indexprotocolname > len(fields) || indexport > len(fields) {
+			return
+		}
+		s := strings.Split(fields[indexport], ":")
+		p, _ := strconv.Atoi(s[1])
+		k.port = p
+		k.protocol = strings.ToUpper(fields[indexprotocolname])
+		k.protocol = strings.ReplaceAll(k.protocol, "\"", "")
+		listeningPorts[k] = ""
+	}
+}
+
+func checkIfListenIsDeclared(listeningPorts, declaredPorts map[key]string) map[key]string {
+	res := make(map[key]string)
+	if len(listeningPorts) == 0 {
+		return res
+	}
+	for k := range listeningPorts {
+		_, ok := declaredPorts[k]
+		if !ok {
+			tnf.ClaimFilePrintf(fmt.Sprintf("The port %d on protocol %s in pod %s is not declared.", k.port, k.protocol, listeningPorts[k]))
+			res[k] = listeningPorts[k]
+		}
+	}
+	return res
+}
+
+//nolint:funlen
+func testListenAndDeclared(env *provider.TestEnvironment) {
+	var k key
+	declaredPorts := make(map[key]string)
+	listeningPorts := make(map[key]string)
+	var failedPods string
+	var skippedPods string
+
+	for _, cut := range env.Containers {
+		ports := cut.Data.Ports
+		fmt.Println(ports)
+		if ports == nil {
+			tnf.ClaimFilePrintf("Failed to get declared port for %s", cut.StringShort())
+			skippedPods += cut.StringShort() + "\n"
+		}
+		for j := 0; j < len(ports); j++ {
+			k.port = int(ports[j].ContainerPort)
+			k.protocol = string(ports[j].Protocol)
+			declaredPorts[k] = ports[j].Name
+		}
+
+		oc := clientsholder.GetClientsHolder()
+		output, outerr, err := oc.ExecCommandContainer(clientsholder.Context{Namespace: cut.Namespace,
+			Podname: cut.Podname, Containername: cut.Data.Name}, `ss -tulwnH`)
+		if err != nil {
+			logrus.Errorln("can't execute command on container ", err)
+			continue
+		}
+		if outerr != "" {
+			logrus.Errorln("error when running listening command ", outerr)
+			continue
+		}
+		parseListening(output, listeningPorts)
+		if len(listeningPorts) == 0 {
+			tnf.ClaimFilePrintf("Failed to get listening port for %s", cut.StringShort())
+			continue
+		}
+		// compare between declaredPort,listeningPort
+		undeclaredPorts := checkIfListenIsDeclared(listeningPorts, declaredPorts)
+		for k := range undeclaredPorts {
+			tnf.ClaimFilePrintf("The port %d on protocol %s not declared on %s", k.port, k.protocol, cut.StringShort())
+		}
+		if len(undeclaredPorts) != 0 {
+			for x := range undeclaredPorts {
+				p := strconv.Itoa(x.port)
+				failedPods += p + " " + x.protocol
+			}
+		}
+	}
+
+	if nf, ns := len(failedPods), len(skippedPods); nf > 0 || ns > 0 {
+		ginkgo.Fail("Found %d pods with listening ports not declared and skipped %d pods due to unexpected error", nf, ns)
```

Comment:
You have a formatted string without the fmt.Sprintf. Also, add a logrus.Debugf to see the whole slice of failed containers. I recommend using this code:
```
if n := len(failedContainers); n > 0 {
    logrus.Debugf("Failed containers: %v", failedContainers)
    ginkgo.Fail(fmt.Sprintf("Found %d pods with listening ports that had not been declared", n"))
}
```

Reply:
**ginkgo.Fail(fmt.Sprintf("Found %d pods with listening ports that had not been declared", n"))** 
about this I'm not sure because failedContainer include all the not declared but part of this is listeining - this come from here- 
	undeclaredPorts := checkIfListenIsDeclared(listeningPorts, declaredPorts)
		for k := range undeclaredPorts {
			tnf.ClaimFilePrintf("The port %d on protocol %s not declared on %s", k.port, k.protocol, cut.StringShort())
		}
		if len(undeclaredPorts) != 0 {
			failedContainers = append(failedContainers, fmt.Sprintf("%s undeclaredPorts: %v", cut.StringShort(), undeclaredPorts))
		}


and have one more type that just only not declared, this come from here -->

	output, outerr, err := oc.ExecCommandContainer(clientsholder.Context{Namespace: cut.Namespace,
			Podname: cut.Podname, Containername: cut.Data.Name}, `ss -tulwnH`)
		if err != nil {
			logrus.Errorln("can't execute command 'ss -tulwnH' on %s, err:%s", cut.StringShort(), err) 
			failedContainers += cut.StringShort() + "\n"
			continue
		}
		if outerr != "" {
			logrus.Errorln("error when running listening command 'ss -tulwnH' on %s, outerr: %s", cut.StringShort(), outerr)
			failedContainers += cut.StringShort() + "\n"
			continue
		}

@greyerof

Reply:
Sorry, when I said "Found %d pods ..." it should be "Found %d containers..."